File: web/WebFirstLab/php/CachedContainable.php
<?php
    require_once('Containable.php');

    class CachedContainable implements Containable {
        private $cache;
        private $containable;

        public function __construct(Containable $containable) {
            $this->containable = $containable;
        }

        public function content() {
            if(empty($this->cache)) {
                $this->cache = $this->containable->content();
            }

            return $this->cache;
        }
    }
?>

File: web/WebSecondLab/src/main/java/app/entities/classes/CachedContainable.java
package app.entities.classes;

import app.entities.exceptions.WrongInputException;
import app.entities.interfaces.Containable;

public class CachedContainable<T> implements Containable<T> {
    private final Containable<T> containable;
    private T cache;

    public CachedContainable(Containable<T> containable) {
        this.containable = containable;
    }

    @Override
    public T content() throws WrongInputException {
        if (cache == null) {
            this.cache = containable.content();
        }

        return this.cache;
    }
}
